enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,239 @@
+  - "@dtantsur"
+reviewers:
+  - "@flaper87"
+approvers:
+  - "@hardys"
+  - "@zaneb"
+creation-date: 2022-03-14
+last-updated: 2022-03-14
+status: implementable
+---
+
+# Fully converge ZTP and Metal3 flows
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes converging two somewhat different bare-metal
+provisioning flows in OpenShift:
+- *Standard* Metal3 flow used in bare-metal IPI and [baremetal-operator][bmo].
+  Uses the [Ironic agent][ipa] (also known as ironic-python-agent or
+  simply IPA) for writing the CoreOS image.
+- ZTP (Zero-Touch Provisioning) flow using the so-called *live ISO* Ironic and
+  baremetal-operator feature, the assisted service and the assisted agent.
+
+## Motivation
+
+Having two different provisioning flows comes with significant maintenance and
+development costs:
+- Some features, namely firmware and RAID configuration, require using the
+  Ironic agent, which is not present in the *live ISO* flow.
+- Supporting (i)PXE boot with the *live ISO* flow requires creating new API
+  in baremetal-operator specifically for this purpose.
+- In the *live ISO* flow, it is not possible to notify Ironic (and thus
+  baremetal-operator) when the actual installation finishes. From the
+  perspective of Metal3 components, the installation is done once the ISO is
+  connected. As a result, Ironic cannot disconnect the ISO from the machine's
+  BMC, nor can it update the BMC's boot device settings properly.
+- The *live ISO* flow disables Ironic *inspection* process and has to simulate
+  it via the assisted components. This creates an unnessesary coupling between
+  the assisted service and the baremetal-operator's hardware inventory format
+  (which, in turn, is derived from the Ironic's inventory format).
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Making assisted components mandatory for all bare-metal deployments.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). It will be updated to use the
+  assisted service's `InfraEnv` resources, when they are present for the given
+  `BareMetalHost`.
```

Comment:
An optional dependency.

Reply:
And if the label is present but the CRD is not installed?

Reply:
Why would a user add an annotation pointing to an InfraEnv on a Host in a cluster that is not running ZTP and has no InfraEnv CRD?
There's no way that any API anywhere can reliably guess what the user meant when they provide garbage input. All we can do is keep attempting to eventually converge, and report any errors in a way that will hopefully make sense to the user.

Reply:
Users do all sorts of crazy things. The controller needs to be able to cope with that case by not crashing, at the very least.

Reply:
> And if the label is present but the CRD is not installed?

I think it will be a normal error propagated through `PreprovisioningImage`, but I can double-check.